### PR TITLE
front: convert RollingStockCardDetail.jsx to tsx and rtk

### LINKED ISSUE #4170: front: convert RollingStockCardDetail.jsx to tsx and rtk
#### Description and goal

Use RTK for endpoints calls

#### Acceptance criteria
 
The andepoints are all called with rtk

### PR DESCRIPTION
**Pull request by Yohh**
close #4170

### CODE REVIEW

**Review by Uriel-Sautron — APPROVED** [review]
Good work !!!

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 1]
```diff
@@ -1,35 +1,51 @@
...
 
-const ROLLING_STOCK_URL = '/editoast/rolling_stock/';
+type RollingStockCardDetailProps = {
+  id: number;
+  curvesComfortList?: string[];
+  setCurvesComfortList: (curvesComfortList: string[]) => void;
+};
 
-export default function RollingStockCardDetail(props) {
+export default function RollingStockCardDetail(props: RollingStockCardDetailProps) {
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
export default function RollingStockCardDetail({ id, curvesComfortList, setCurvesComfortList } : RollingStockCardDetailProps) {
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 2]
```diff
@@ -40,34 +56,42 @@ export default function RollingStockCardDetail(props) {
       };
       // AC & HEATING curves (optional)
       curvesData.modes[mode].curves.forEach((curve) => {
-        const serieId = `${name} ${curve.cond.comfort}`;
-        transformedCurves[serieId] = { ...curve.curve, mode: name, comfort: curve.cond.comfort };
+        if (curve.cond) {
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
        if (curve.cond?.comfort) {
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 3]
```diff
@@ -40,34 +56,42 @@ export default function RollingStockCardDetail(props) {
...
+          const serieId = `${name} ${curve.cond.comfort}`;
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
          const optionalCurveName = `${name} ${curve.cond.comfort}`;
```
Can you rename this variable ? (suggestion)

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 4]
```diff
@@ -1,35 +1,51 @@
...
-  const mode2name = (mode) => (mode !== 'thermal' ? `${mode}V` : t('thermal'));
+  const mode2name = (mode: string) => (mode !== 'thermal' ? `${mode}V` : t('thermal'));
 
-  const listCurvesComfort = (curvesData) => {
-    const list = ['STANDARD'];
-    Object.keys(curvesData.modes).forEach((mode) => {
-      curvesData.modes[mode].curves.forEach((curve) => {
-        if (!list.includes(curve.cond.comfort)) list.push(curve.cond.comfort);
+  const listCurvesComfort = (curvesData: RollingStock['effort_curves']) => {
+    if (curvesData) {
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
```
No need for this if thanks to Typescript :)

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 5]
```diff
@@ -1,35 +1,51 @@
...
+      const list = ['STANDARD'];
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
      const comfortList = ['STANDARD'];
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 6]
```diff
@@ -1,35 +1,51 @@
...
+      Object.keys(curvesData.modes).forEach((mode) => {
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: EffortCurves are always a bit hard to manipulate, I think it's better to add some variables to clarify what we are doing, like this :
```suggestion
    const rollingStockModes = Object.keys(curvesData.modes);
    rollingStockModes.forEach((mode) => {
```
What do you think ?

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 7]
```diff
@@ -1,35 +1,51 @@
...
+        curvesData.modes[mode].curves.forEach((curve) => {
+          if (curve.cond) {
+            if (!list.includes(curve.cond.comfort as string))
+              list.push(curve.cond.comfort as string);
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
          if (curve.cond?.comfort) {
            if (!list.includes(curve.cond.comfort))
              list.push(curve.cond.comfort);
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 8]
```diff
@@ -40,34 +56,42 @@ export default function RollingStockCardDetail(props) {
...
-    try {
-      const rollingStockData = await get(`${ROLLING_STOCK_URL + id}/`);
-      setData(rollingStockData);
-      setCurvesComfortList(listCurvesComfort(rollingStockData.effort_curves));
-    } catch (e) {
+  const { data, error } = osrdEditoastApi.useGetRollingStockByIdQuery(
+    { id },
+    {
+      skip: !id,
+      refetchOnMountOrArgChange: true,
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: Why this refetchOnMountOrArgChange is needed ?

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 9]
```diff
@@ -40,34 +56,42 @@ export default function RollingStockCardDetail(props) {
...
     return transformedCurves;
   };
 
-  const getRollingStock = async () => {
-    try {
-      const rollingStockData = await get(`${ROLLING_STOCK_URL + id}/`);
-      setData(rollingStockData);
-      setCurvesComfortList(listCurvesComfort(rollingStockData.effort_curves));
-    } catch (e) {
+  const { data, error } = osrdEditoastApi.useGetRollingStockByIdQuery(
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
  const { data: rollingStock, error } = osrdEditoastApi.useGetRollingStockByIdQuery(
```
You can rename the result of the api this way ;)

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 10]
```diff
@@ -40,34 +56,42 @@ export default function RollingStockCardDetail(props) {
...
+  }, [data]);
+
+  useEffect(() => {
+    if (error) {
+      const err = error as ApiError;
       dispatch(
         setFailure({
           name: t('osrdconf:errorMessages.unableToRetrieveRollingStock'),
-          message: e.message,
+          message: err.data.message,
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
          message: (error as ApiError).data.message,
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 11]
```diff
@@ -103,7 +127,7 @@ export default function RollingStockCardDetail(props) {
               <tr>
                 <td className="text-primary">{t('timetableGamma')}</td>
                 <td>
-                  {data.gamma.value * -1}
+                  {data.gamma?.value ? data.gamma.value * -1 : null}
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
                  {data.gamma.value * -1 }
```
Update the openApi.yaml to make the gamma and gamma.value not undefined

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 12]
```diff
@@ -172,7 +202,7 @@ export default function RollingStockCardDetail(props) {
         </div>
       </div>
       <RollingStockCurves
-        data={transformCurves(data.effort_curves)}
+        data={transformCurves(data.effort_curves as RollingStock['effort_curves'])}
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]:
```suggestion
        data={transformCurves(data.effort_curves)}
```
Not needed here

**Review by clarani — CHANGES_REQUESTED** [review]
Nice job !! 
It would be nice to modify the openApi.yaml to avoid to test everywhere if a value is undefined or not.

Also, can you separate your changes (migrate RollingStockCardDetail to ts in this PR, rename isStdcm into isCondensed in another one) ? Thanks !

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 8 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: to refetch if we change the pathfinding, but without a useEffect, this way we limitate the re-renders

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 1 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 2 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 3 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 5 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 4 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 6 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: need to talk about that

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 7 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 9 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 10 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 11 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: need to talk about that

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 12 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 11 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 6 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

File: front/src/styles/scss/applications/_stdcm.scss [hunk 13]
```diff
@@ -7,6 +7,6 @@
   }
 }
 
-.stdcm-allowance-icon {
```
Comment (clarani) [inline front/src/styles/scss/applications/_stdcm.scss]: You forgot this in this PR :)

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 14]
```diff
@@ -40,36 +55,42 @@ export default function RollingStockCardDetail(props) {
...
-      const rollingStockData = await get(`${ROLLING_STOCK_URL + id}/`);
-      setData(rollingStockData);
-      setCurvesComfortList(listCurvesComfort(rollingStockData.effort_curves));
-    } catch (e) {
+  const { data: rollingStock, error } = osrdEditoastApi.useGetRollingStockByIdQuery(
+    { id },
+    {
+      skip: !id,
+    }
+  );
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx:76]: Can you add this comment ? :) 
```suggestion
  // we only fetch the whole rollingStock here, when we open the card and display its details
  const { data: rollingStock, error } = osrdEditoastApi.useGetRollingStockByIdQuery(
    { id },
    {
      skip: !id,
    }
  );
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 15]
```diff
@@ -1,35 +1,50 @@
...
+  id: number;
+  curvesComfortList?: string[];
+  setCurvesComfortList: (curvesComfortList: string[]) => void;
+};
 
-export default function RollingStockCardDetail(props) {
+export default function RollingStockCardDetail({
+  id,
+  curvesComfortList,
+  setCurvesComfortList,
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx:19]:
```suggestion
  // TODO: the curvesComfortList should be created in the parent component
  curvesComfortList,
  setCurvesComfortList,
```

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx [hunk 16]
```diff
@@ -1,35 +1,50 @@
...
+        }
       });
     });
-    return list;
+    return comfortList;
   };
 
-  const transformCurves = (curvesData) => {
-    const transformedCurves = {};
+  const transformCurves = (curvesData: RollingStock['effort_curves']) => {
```
Comment (clarani) [inline front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: Can you modify the function so it only take the modes in argument ?
```suggestion
  const transformCurves = (rollingStockCurves: RollingStock['effort_curves']['modes']) => {
```

**Review by clarani — CHANGES_REQUESTED** [review]
3 small comments :)

File: front/src/styles/scss/applications/_stdcm.scss (hunk 13 above)
Comment (Yohh) [reply front/src/styles/scss/applications/_stdcm.scss]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 14 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx:76]: done

File: front/src/common/RollingStockSelector/RollingStockCardDetail.tsx (hunk 16 above)
Comment (Yohh) [reply front/src/common/RollingStockSelector/RollingStockCardDetail.tsx]: done

**Review by clarani — APPROVED** [review]
LGTM ! Tested on local :)

**Pull request merged**